IdentityCore/src/throttling/thumbprint_calculation/MSIDSilentRequestThumbprintCalculator.h

```diff
@@ -0,0 +1,40 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Foundation/Foundation.h>
+#import "MSIDThumbprintCalculatable.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDSilentRequestThumbprintCalculator : NSObject <MSIDThumbprintCalculatable>
```

Comment:
Original comments: https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/907#discussion_r540670865
**@oldalton commented:** Do we need to have a separate thumbprint calculator for each request type? Could we instead demo something more generic - e.g. have request expose parameters that should be included in thumbprint? Otherwise there seems to be quite a bit of code duplication.
**@NerevarineRule replied:** 
The challenge is that for non-SSO requests, we gather authority realm, environment, and home account id from MSIDRequestParameters, and the remaining request parameters from MSIDTokenRequest type.

For the SSO request type, we query from the JSON dictionary that gets information from MSIDConfiguration, Authority, MSIDBrokerOperationTokenRequest

Some of the parameters are optional - auth scheme & extra query URL.

So, wouldn't it be best to abstract the difference between the two request types by creating a common protocol that declares thumbprint calculation API, then we will have a wrapper class for non-SSO request and one for SSO request. They will have their own local NSmutableDictionary, which will be initialized from the locations that will include all the parameters we will need for full request thumbprint calculation + some extra parameters.
Then,

Initialize thumbprint calculator class with all request parameters that are potential candidates
use white list for strict request thumbprint to only use the parameters in the white list set (since there are only 5 parameters for strict request thumbprint vs ~20 potential candidates)
use black list for full request thumbprint to exclude parameters from full request thumbprint calculation. (ex: broker_key, client_version, etc..)
Given that SSO request vs non-SSO request have different request parameters that use different key names that are queried from multiple different locations and they have their own set of optional parameters (broker key for SSO request, for instance) above approach seems to make more sense.

Then Hieu's throttling decision handler can have an API that accepts input type of id to calculate thumbprint and then uses LRU cache class.

Hieu's throttling handler can initialize MSIDSilentRequestThumbprintCalculator class in MSIDSilentRequest - refreshTokenRequestWithRequestParameters using MSIDTokenRequest parameters & little bit of MSIDrequestParameters. then pass it into his API.

Likewise, his throttling handler can intialize MSIDSSOSilentRequestThumbprintCalculator class using MSIDBrokerOperationSilentTokenRequest's json Dictionary in the SSORequest executeRequestWithCompletion.

Then his throttling handler can simply call "getFullRequestThumbprint" to get the thumbprint key for the UI Required LRU Cache, and "getStrictRequestThumbprint" to get the thumbprint key for the 429/5xx (Server Delay Cache) without having to know whether the originating request was of SSO extension type or non-SSO type.

Does that help to answer your question? Otherwise, could you share your thoughts in the Teams channel?